src/nccl_ofi_net.c

```diff
@@ -1414,40 +1416,73 @@ static ncclResult_t ofi_connect(int dev, void *handle, void **sendComm)
 	}
 
 	/* Send "connect" message to remote EP */
-	do {
-		/*
-		 * TODO: replace it with API of FI_INJECT type when most of
-		 * providers can support it, so that need for completion check
-		 * below can be lifted.
-		 */
-		rc = fi_tsend(sComm->local_ep, (void *)&local_ep_addr,
-			      MAX_EP_ADDR, NULL, sComm->remote_ep,
-			      sComm->tag | ~max_tag, &req->ctx);
-		if (rc == 0)
-			break;
+	/* do not wait for completion in case of sending to self */
+	if (memcmp(local_ep_addr, remote_ep_addr, namelen) == 0) {
+		static char buffer_to_self[MAX_EP_ADDR];
+		NCCL_OFI_TRACE(NCCL_NET, "Connect to self");
+		if (request_to_self != NULL) {
+			NCCL_OFI_WARN("Another connect to self request already in progress");
+			ret = ncclSystemError;
+			goto error;
+		}
+		memcpy(buffer_to_self, local_ep_addr, MAX_EP_ADDR);
```

Comment:
What is the need to use `buffer_to_self`? Couldn't we directly use `local_ep_addr` below?

Reply:
local_ep_addr is on stack, we need memory to be accessible _after_ this function returns.

Reply:
Ohh, since you aren't processing completion, the application buffer needs to stay and hence the copy. I would appreciate a comment explaining that.

Reply:
Reworked